Sources/SourceKitLSP/Swift/SyntaxTreeManager.swift

```diff
@@ -0,0 +1,112 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftParser
+import SwiftSyntax
+
+/// Keeps track of SwiftSyntax trees for document snapshots and computes the
+/// SwiftSyntax trees on demand.
+actor SyntaxTreeManager {
+  /// A task that parses a SwiftSyntax tree from a source file, producing both
+  /// the syntax tree and the lookahead ranges that are needed for a subsequent
+  /// incremental parse.
+  private typealias SyntaxTreeComputation = Task<(tree: SourceFileSyntax, lookaheadRanges: LookaheadRanges), Never>
```

Comment:
Not really a comment on this PR, but is there a reason `Parser.parseIncrementally` returns a tuple instead of a dedicated struct? Seems like a single type would be easier to manage, and would help ensure you don't pass a mismatching syntax tree + lookahead ranges to `IncrementalParseTransition` (and would make it trivial to add any additional state later down the road if needed).

Reply:
Yeah, I thought the same when using the API. We should probably change that

Filed https://github.com/apple/swift-syntax/issues/2267